java/src/org/openqa/selenium/netty/server/ResponseConverter.java

```diff
@@ -59,7 +60,7 @@ public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise)
     // We may not know how large the response is, but figure it out if we can.
     byte[] ary = new byte[CHUNK_SIZE];
     InputStream is = seResponse.getContent().get();
-    int byteCount = is.read(ary);
+    int byteCount = ByteStreams.read(is, ary, 0, ary.length);
```

Comment:
I do not understand why this is needed. I thought `is.read(ary)` becomes `is.read(ary, 0, ary.length)`.

From the PR description, I understood that the initial 1MB chunck is read correctly, right?

Reply:
This is not related to the chunking, it is more about the correct usage of the InputStream.

The InputStream.read method does not guarantee to fill the complete array, even if there is more data in the stream. The InputStream.read javadoc states: 
`Reads some number of bytes from the input stream and stores them into the buffer array b. The number of bytes actually read is returned as an integer. This method blocks until input data is available, end of file is detected, or an exception is thrown.`

An InputStream could - for whatever reason - only read some bytes from the stream and write it into the array. This would fullfill the guarantees of the javadoc. The implementation here would assume the end of stream is reached and close the stream.

On the other hand the ByteStream.read javadoc: 
`Reads some bytes from an input stream and stores them into the buffer array b. This method blocks until len bytes of input data have been read into the array, or end of file is detected. `

There is a guarantee the array is filled to the end or the end of stream is reached. This allows detect the EOF by checking the number of bytes read, what is done in the current implementation.

Best wishes
Jörg

Reply:
PS: is.readNBytes(ary, 0, ary.length) is Java 9 and i think Selenium is at Java 8.